README.md
![Latest Release Download Count](https://img.shields.io/badge/dynamic/json?label=Downloads@latest&query=assets%5B1%5D.download_count&url=https%3A%2F%2Fapi.github.com%2Frepos%2Fspetzel2020%2Fquick-encounters%2Freleases%2Flatest)
![Forge Installs](https://img.shields.io/badge/dynamic/json?label=Forge%20Installs&query=package.installs&suffix=%25&url=https%3A%2F%2Fforge-vtt.com%2Fapi%2Fbazaar%2Fpackage%2Fquick-encounters&colorB=4aa94a)

# Quick Encounters
## 1.2
* New feature: [#123](https://github.com/spetzel2020/quick-encounters/issues/123): New option to automatically add Player Tokens to the Combat Tracker
* Compatible with Foundry v9 - v11
## 1.1
* New features: Default folder for Quick Encounters (Issue #94), Add Actor RollTables from Journal Entries (Issue #40) - see new Settings
* Bugs fixed: See [Release Notes](https://github.com/spetzel2020/quick-encounters/blob/master/CHANGELOG.md)

## 1.0
* v1.0.8 now working with Foundry v9 & v10
* Bugs fixed: See [Release Notes](https://github.com/spetzel2020/quick-encounters/blob/master/CHANGELOG.md)
* Instant Encounters: Enable the new Setting, then you can drag your Quick Encounter directly from Journals to the Scene and be prompted to run it immediately! Perfect for Random or Wandering Encounters.

* **Author**: Spetzel#0103
* **Version**: 1.2.1 
* **Foundry VTT Compatibility**: 9-11
* **System Compatibility (If applicable)**: All; additional features for dnd5e
* **Translation Support**: en, es (thanks [lozalojo!](https://github.com/lozalojo)), ja (thanks "touge"!), de (thanks @Fallayn#6414), it (thanks [riccisi](https://github.com/riccisi))


# Description
Keeps your Scene map clean and your encounters simple by saving tokens/Actors into Quick Encounter Journal Entries:
* Design your encounters by positioning and configuring Hostile tokens, then save the tokens into a Quick Encounter Journal Entry, marked with a single Map Note on the Scene which only you the GM see
* Run your Quick Encounter by clicking the "Run Quick Encounter" button; all your saved tokens will be positioned back on the map and added to the Combat Tracker, ready to fight!
* Run encounters from converted or packaged modules with embedded Actors
* Easily scale-up encounters by changing "3 Zombies" to "10 Zombies"
* Similar to the Encounter ability in Fantasy Grounds
* For Roving Encounters move a single Map Note, not many tokens
* Save your player tokens at the entrance to a new Scene too!

Quick Encounters works well with [TokenMold](https://github.com/Moerill/token-mold#token-mold) to vary your tokens names, HP, etc (although TokenMold is not required for Quick Encounters). Quick Encounters also previews the XP for each group of hostiles if using dnd5e.

# Install
1. Go to the "Add-on Modules" tab in Foundry Setup
2. Click "Install Module" and search for **Quick Encounters** OR paste this link: "https://github.com/spetzel2020/quick-encounters/releases/latest/download/module.json"
3. Open your world and go to Settings>Manage Modules and enable Quick Encounters

# Using Quick Encounters
## Method 1 (like Fantasy Grounds "Encounters")
1. Drag your Hostile tokens (or now Tiles!) to the map and select them
2. Then click the Quick Encounters button (in the Token Controls palette on the left); a new Journal Entry will be created with your tokens/tiles and their positions saved in it
3. Close the Journal Entry and you will see a new Quick Encounter Map Note on the Scene; double-right-click to edit the icon, color, etc.
4. To run your Quick Encounter, simply double-click the map Note and click the "Add to Combat Tracker" button; all your saved tokens/tiles will be positioned on the map and added to the Combat Tracker, ready to fight!

## Method 2
1. Drag Actors into an existing or new Journal Entry describing the encounter (you can add how many of them you want).
2. Close the Journal Entry
3. Now drag the Journal Entry to the Scene as a Note at the desired encounter position.
4. To run your Quick Encounter, simply double-click the map Note to open the Quick Encounter Journal Entry.
5. Click the Quick Encounter button: your saved tokens will be positioned around the map Note, and added to the Combat Tracker.

### Using Compendium links in Method 2
* In Step 1, you can select Actors from a Compendium and drag them directly into the Journal Entry. This lets you defer importing Actors until you actually need them (for example for rare Encounters). Currently you need to manually remove the imported Actors when and if you run the Quick Encounter.
* Note that you cannot drag the Compendium Actor to the Scene to place tokens, because that will automatically import the Actor and the link will then be to the imported Actor, not to the Compendium.

## Method 3 (save tokens/tiles into existing Journal Entries)
If you have a licensed module with existing Journal Entries you probably don't want to edit all of them to insert Actors (Method 2). Using the new Quick Encounter Dialog instead:
1. Create Hostile tokens (representing the Encounter) and drag them to the Scene
2. Open the Journal Entry you want to associate with the Quick Encounter
3. Click the Quick Encounter button and you will be asked if you want to create a new Encounter (Method 1) or add the tokens to this Journal Entry
4. To run the Encounter later, open the Journal Entry and use the Run Quick Encounter button in the Quick Encounter Dialog

## Working with Tokens and Tiles
If you have BOTH Tokens and Tiles you want in your Quick Encounter, create the Quick Encounter with the tokens (and one of the above methods). Leave the Quick Encounter open and switch to the Tile Controls menu (left side-bar). Place and select your tokens and click the Quick Encounter button to add the tiles to the open Quick Encounter.

Coming soon: Drag-and-drop for tokens, tiles, and other stuff!

## The Quick Encounter dialog
![QuickEncounterDialog](https://github.com/spetzel2020/quick-encounters/blob/master/img/CompanionDialog.png)

## Contributions
[![ko-fi](https://www.ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/T6T82XFQD)

## License
**Quick Encounters for Foundry VTT** by Jeffrey Pugh is licensed under the [GNU General Public License v3.0](https://github.com/spetzel2020/quick-encounters/blob/master/LICENSE)

This work is licensed under Foundry Virtual Tabletop [EULA - Limited License Agreement for module development v 0.1.6](http://foundryvtt.com/pages/license.html).


CHANGELOG.md
# RELEASE NOTES
## 1.2
- New feature: [#123](https://github.com/spetzel2020/quick-encounters/issues/123): New option to automatically add Player Tokens to the Combat Tracker
### 1.2.2
- Fixing [#137](https://github.com/spetzel2020/quick-encounters/issues/137) (support more generalized dice rolls)
### 1.2.1
- Fixed [#139](https://github.com/spetzel2020/quick-encounters/issues/139) (filter for playerToken.actor === user.character)
## 1.1
- New features: Add a default folder for new QEs: [#94](https://github.com/spetzel2020/quick-encounters/issues/94), Add Actor RollTables from Journal Entries: [#40](https://github.com/spetzel2020/quick-encounters/issues/40) 
- Fixed [#117](https://github.com/spetzel2020/quick-encounters/issues/117): Missing i18n tags
- Fixed [#118](https://github.com/spetzel2020/quick-encounters/issues/118)
- Fixed [#70](https://github.com/spetzel2020/quick-encounters/issues/70)
### 1.1.6
- Fixed [#119] (https://github.com/spetzel2020/quick-encounters/issues/119)
### 1.1.5
- Fixed [#83] (https://github.com/spetzel2020/quick-encounters/issues/83)
- Fixed [#131] (https://github.com/spetzel2020/quick-encounters/issues/131)
- Fixed [#132] (https://github.com/spetzel2020/quick-encounters/issues/132)
- Fixed [#134] (https://github.com/spetzel2020/quick-encounters/issues/134)
- Fixed [#135] (https://github.com/spetzel2020/quick-encounters/issues/135)
### 1.1.4
- Fixed [#115] (https://github.com/spetzel2020/quick-encounters/issues/115)
- Fixed [#122] (https://github.com/spetzel2020/quick-encounters/issues/122)
- Fixed [#130] (https://github.com/spetzel2020/quick-encounters/issues/130)
### 1.1.3
- Fixed Issues [#126](https://github.com/spetzel2020/quick-encounters/issues/126), [#127](https://github.com/spetzel2020/quick-encounters/issues/127)
- Fixed Issue [#105](https://github.com/spetzel2020/quick-encounters/issues/105)

### 1.1.2
- Fixed Issue [#121](https://github.com/spetzel2020/quick-encounters/issues/121)
- Issue [#120](https://github.com/spetzel2020/quick-encounters/issues/120): Fixed icon not dimming
### 1.1.1
- Fixed Issue [#40](https://github.com/spetzel2020/quick-encounters/issues/40)
### 1.1.0
- Fixed [#94](https://github.com/spetzel2020/quick-encounters/issues/94): Add a default folder for new QEs
- Fixed [#108](https://github.com/spetzel2020/quick-encounters/issues/108): Removed Hide button in QE because of multi-page Journals and always add Show button (for when you have Show QE Automatically set false)
- Fixed [#114](https://github.com/spetzel2020/quick-encounters/issues/114): QE dialog didn't appear when you created a new QE
- Fixed [#116](https://github.com/spetzel2020/quick-encounters/issues/116): Compendium entries aren't working
## 1.0
- Issue #71: Instant Encounters: Drag a Quick Encounter directly from the Journals directory to the Scene to run it!
### 1.0.8
-  Fixed [#112](https://github.com/spetzel2020/quick-encounters/issues/112): More references to JournalEntryPage
### 1.0.7
- Fixed [#109](https://github.com/spetzel2020/quick-encounters/issues/109): Critical bug with Foundry v9
### 1.0.5
- Fixed: [#103](https://github.com/spetzel2020/quick-encounters/issues/103); removed support for clicking "fist" icon for an open QE since there can now be multiple open (with multi-page Journals)
- Fixed: [#98](https://github.com/spetzel2020/quick-encounters/issues/98)
- Fixed: [#99](https://github.com/spetzel2020/quick-encounters/issues/99)
- Fixed: [#104](https://github.com/spetzel2020/quick-encounters/issues/104)
- Fixed: [#106](https://github.com/spetzel2020/quick-encounters/issues/106)
### 1.0.4
- Updated Spanish translation (thanks https://github.com/lozalojo!)
- Testing with Foundry v10 Testing 1 Release (v10.275) - fatal bugs
- Testing with Foundry v10 Testing 3 Release (v10.277)
- Fixed: #91 Error running Quick Encounter with Foundry 10 Testing 1 (v10.275)
- Fixed: Compatibility warnings and other Foundry v10-related bugs
### 1.0.3
Basic compatibility with Foundry v10
- When extracting embedded Actors in JE HTML use class="content-link" instead of class="entity-link"
- Fixed: Quick Encounter button now works (basic testing)

### 1.0.2
- Fixed: posion.svg -> poison.svg typo
- Fixed: Issue [#88](https://github.com/spetzel2020/quick-encounters/issues/88)
## 0.9 
- Supports Foundry 0.8+ and 9.x only
### 0.9.10
- Fixed: Issue #81 (Run Encounter from Compendium fails) - fixed with PR#80 (thanks https://github.com/jsabol )

modules/Globals.js
"use strict";
// Created from Globals.swift
//  8/17/2020   Created/Copied

//Singleton class for globally managing detail output
export const Output =  {
    _isEnabled : false,
    _output : "",

    add(...strings) {
      this._output += strings.join('');
    },

    initialize() {
      this._output = "";
      this._isEnabled = true;
    },

    get output() {
      return this._output;
    }
}

export var printError = function(error, explicit) {
    console.log(`[${explicit ? 'EXPLICIT' : 'INEXPLICIT'}] ${error.name}: ${error.message}`);
}

export async function loadReferenceData(...files) {
  var spellReference = [];
  // Load external JSON spell data files
  for (let file of files)  {
    try {
        const spellsJSON = await fetch(file);
        //Format of spells JSON is an embedded array
        const spellsData = await spellsJSON.json();
        //Add all the elements of spells to spellReference
        spellReference = spellReference.concat(spellsData.spells);
    } catch (e) {
        if (e instanceof SyntaxError) {
            printError(e, true);
        } else {
            printError(e, false);
        }
    }
  }
  return spellReference;
}


modules/EncounterNote.js
import {QuickEncounter, QE, Dialog3} from './QuickEncounter.js';
/*
Extend the placeable Map Note - select the desired tokens and then tap the Quick Encounters button
Subsequently can add: (a) Drag additional tokens in, (b) populate the Combat Tracker when you open the note?
27-Aug-2020   Created
30-Aug-2020   Added EncounterNoteConfig
13-Sep-2020    QuickEncounter.deleteNote moved/renamed to EncounterNote.delete
                QuickEncounter.placeNote moved/renamed to EncounterNote.place
                Fixed flow of deleteJournalEntry --> delete associated Note
14-Sep-2020     Display simple dialog when you delete the Map Note corresponding to a Quick Encounter Journal Entry
15-Sep-2020     v0.4.0 i18n for deleting Journal Note
                v0.4.1 delete() - rewrite for getEncounterScene returning the scene not the ID
16-Sep-2020     v0.4.1 place() - if there aren't token coords, and option=placeDefault, then place a map note in the center
21-Sep-2020     v0.4.2: BUG: Dialog.prompt doesn't exist in Foundry 0.6.6 - replace with our own
26-Sep-2020     v0.5.0: Use QuickEncounter.switchToMapNoteScene
27-Sep-2020     v0.5.0: Bypass the Note Config sheet - just create it and allow for updating later
                v0.5.0: NYI: Base code for Hook on renderNoteCOnfig to change it to look like a QE Note (but need a way of determining a QE Note)
9-Nov-2020      v0.6.1: Refactor Map Note related functions here: switchToMapNoteScene(), noMapNoteDialog(), mapNoteIsPlaced()
15-Nov-2020     v0.6.2: Refactor to use quickEncounter rather than journalEntry
2-Dec-2020      v0.6.12: Test parameterization of i18n strings, using Localization.format() (but be backward compatible)
7-Dec-2020      v0.6.13: delete(): Delete ALL of the Notes (in this or other Scenes) associated with the delete journal Entry
                        EncounterNote.create() and .place() return newNote so that we can store that in the quickEncounter before serialization
                        Expand the available list of Note icons (perhaps would be good to have a Setting to allow/disallow this)
8-Aug-2021      0.8.3a: If Foundryv8 then don't do Note deletion in the current scene because new Journal._onDelete() method takes care of that  
                        (although because of https://gitlab.com/foundrynet/foundryvtt/-/issues/5700 this won't work at all currently)  
15-Nov-2021     v0.9.1b: Issue #57 Reintroduce deleltion of notes; doesn't seem to be handled in Foundry 0.8.9     
6-Dec-2021      0.9.3a: Check for Foundry 0.9 OR 0.8   
15-Dec-2021     0.9.3f: EncounterNote.create(): Check/fix deprecation warning by using canvas.scene.embeddedDocuments()      
21-Dec-2021     0.9.5a: Use QuickEncounter.isFoundryV8Plus test   
17-Feb-2022     1.0.1a: Add Hook on dropCanvasData to intercept creation of Notes for Quick Encounters      
26-Feb-2022     1.0.1d: checkForInstantEncounter(): pass new options variable with IE information to QuickEncounter.run()  
3-Mar-2022     1.0.1f: Fixed #84 delete(): corrected to use deleteEmbeddedDocuments()              
4-May-2022      1.0.2b: Fixed: Misspelled poison.svg in moreNoteIcons
31-Aug-2022     1.0.4k: Get parent JournalEntry for JournalEntryPage (because Notes are associated with Journal Entries)
1-Sep-2022      1.0.4k: Move getEncounterScene() to EncounterNote from QuickEncounter
                Check journalEntry to see if we should be looking at parent\
                Also in mapNoteIsPlaced
                1.0.4l: place(): Use canvas.stage.hitArea in v10
                delete(): More deprecation warnings
                create(): Typo in entryId
                1.0.4m: dropCanvasData Hook: Check first JournalEntryPage for a QE (FIX: Should check all)
3-Oct-2022      1.0.7a: Issue #109: Check for FoundryV10 when referencing journalEntryPage or journalEntry.parent   
6-Oct-2022      1.0.8a: Issue #112: Further references to JournalEntryPage    
25-Apr-2023     1.1.5a: Issue #132: Intercept drag of JEPage just like Journal Entries (replace if it has a QE in it)
                Hooks.on('dropCanvasData'): Add check for JournalEntryPage and add common function checkForQEAndCreateNote()
                create(): Remove check for FoundryV10 and switch to parentJournalEntry
                Re-clone _onDropData() from foundry.js to account for new Note fields (for JournalEntryPage)
29-May-2023     1.1.5b: Changed isFoundryV10Plus to isFoundryV10PlusPlus (to support checks for Foundry V11)                  
*/

//Expand the available list of Note icons
const moreNoteIcons = {
    "Acid" : "icons/svg/acid.svg",
    "Angel" : "icons/svg/angel.svg",
    "Aura" : "icons/svg/aura.svg",
    "Blind" : "icons/svg/blind.svg",
    "Blood" : "icons/svg/blood.svg",
    "Bones" : "icons/svg/bones.svg",
    "Circle" : "icons/svg/circle.svg",
    "Clockwork" : "icons/svg/clockwork.svg",
    "Combat" : "icons/svg/combat.svg",
    "Cowled" : "icons/svg/cowled.svg",
    "Daze" : "icons/svg/daze.svg",
    "Deaf" : "icons/svg/deaf.svg",
    "Direction" : "icons/svg/direction.svg",
    "Door-Closed" : "icons/svg/door-closed.svg",   
    "Door-Exit" : "icons/svg/door-exit.svg",    
    "Down" : "icons/svg/down.svg",
    "Explosion" : "icons/svg/explosion.svg",
    "Eye" : "icons/svg/eye.svg",
    "Falling" : "icons/svg/falling.svg",    
    "Frozen" : "icons/svg/frozen.svg",
    "Hazard" : "icons/svg/hazard.svg",
    "Heal" : "icons/svg/heal.svg",
    "Holy Shield" : "icons/svg/holy-shield.svg",
    "Ice Aura" : "icons/svg/ice-aura.svg",
    "Lightning" : "icons/svg/lightning.svg",
    "Net" : "icons/svg/net.svg",
    "Padlock" : "icons/svg/padlock.svg",   
    "Paralysis" : "icons/svg/paralysis.svg",
    "Poison" : "icons/svg/poison.svg",
    "Radiation" : "icons/svg/radiation.svg",
    "Sleep" : "icons/svg/sleep.svg",
    "Sound" : "icons/svg/sound.svg",  
    "Sun" : "icons/svg/sun.svg",
    "Terror" : "icons/svg/terror.svg",   
    "Up" : "icons/svg/up.svg",
    "Wing" : "icons/svg/wing.svg"      
}
Object.assign(CONFIG.JournalEntry.noteIcons, moreNoteIcons);



export class EncounterNoteConfig extends NoteConfig {
    /** @override  */
    //WARNING: Do not add submitOnClose=true because that will create a submit loop
    static get defaultOptions() {
        return mergeObject(super.defaultOptions, {
            id : "encounter-note-config",
            title : game.i18n.localize( "QE.Config.TITLE")
        });
    }
}

export class EncounterNote {
    static async create(quickEncounter, noteAnchor) {
        if (!quickEncounter) {return;}

        const journalEntry = quickEncounter.journalEntry;
        //1.0.4k: REMOVED IN 1.1.5: Use parent (which is what is saved to the map) if this is JournalEntryPage
        //1.0.7a: Check for FoundryV10
        //1.1.5a: Remove defaulting to parent, but if it's a page you have to reference the parent JournalEntry and page reference
        let noteJournalEntry = (journalEntry instanceof JournalEntryPage) ? journalEntry.parent : journalEntry;
        // Create Note data
        const noteData = {
              entryId: noteJournalEntry.id,
              x: noteAnchor.x,
              y: noteAnchor.y,
              icon: CONFIG.JournalEntry.noteIcons.Combat,
              iconSize: 80,
              iconTint: "#FF0000",  //Red
              //Don't specify the name so it inherits from the Journal
              textAnchor: CONST.TEXT_ANCHOR_POINTS.TOP,
              fontSize: 24
        };
        if (journalEntry instanceof JournalEntryPage) {noteData.pageId = journalEntry.id;}

        //v0.5.0: Switch to Note.create() to bypass the Note dialog
        //This is different from the JournalEntry._onDropData approach
        //0.9.3f: Remove deprecation warning by using createEmbeddedDocuments()
        let newNote = QuickEncounter.isFoundryV8Plus ? await canvas.scene.createEmbeddedDocuments("Note",[noteData]) : await Note.create(noteData);
        //1.0.2c: createEmbeddedDocuments returns an array, and we just want a single element
        if (Array.isArray(newNote)) {newNote = newNote[0];}
        newNote._sheet = new EncounterNoteConfig(newNote);
        return newNote;
    }

    static async delete(journalEntry) {
        //1.0.4k: This should always be a real parent JournalEntry
        
        if (!game.user.isGM) {return;}
        //Create filtered array of matching Notes for each scene
        let matchingNoteIds;
        let numNotesDeleted = 0;
        for (const scene of game.scenes) {
            if (QuickEncounter.isFoundryV10Plus) {
                matchingNoteIds = Array.from(scene.notes?.values()).filter(nd => nd.entryId === journalEntry.id).map(note => note.id);
            } else if (QuickEncounter.isFoundryV8Plus) {
                matchingNoteIds = Array.from(scene.data.notes.values()).filter(nd => nd.data.entryId === journalEntry.id).map(note => note.id);
            } else {
                matchingNoteIds = scene.data.notes.filter(nd => nd.entryId === journalEntry.id).map(note => note._id);
            }
            if (!matchingNoteIds?.length) {continue;}
            //Deletion is triggered by Scene (because that's where the notes are stored)
            //v0.8.3a: If Foundry v0.8.x then don't delete the Note in the viewed Scene because the Journal._onDelete() trigger does that
            //v0.9.1b: Issue #57 Reintroduce deleltion of notes; doesn't seem to be handled in Foundry 0.8.9
            scene.deleteEmbeddedDocuments("Note", matchingNoteIds);
            numNotesDeleted += matchingNoteIds.length;
        }

        if (numNotesDeleted) {
            //0.4.2: Replaces Dialog.prompt from Foundry 0.7.2
            EncounterNote.dialogPrompt({
                title: game.i18n.localize("QE.DeletedJournalNote.TITLE"),
                content: game.i18n.format("QE.DeletedJournalNote.Multiple.CONTENT",{numNotesDeleted}),
                label : "",
                callback : () => {console.log(`Deleted ${numNotesDeleted} Map Note(s)`);},
                options: {
                top:  window.innerHeight - 350,
                left: window.innerWidth - 720,
                width: 400,
                jQuery: false
                }
            });
        }

    }

    static dialogPrompt({title, content, label, callback}={}, options={}) {
        return new Promise(resolve => {
          const dialog = new Dialog({
            title: title,
            content: content,
            buttons: {
              close: {
                icon: '<i class="fas fa-check"></i>',
                label: label,
                callback: callback
              }
            },
            default: "close",
            close: resolve
          }, options);
          dialog.render(true);
        });
    }

    static async place(quickEncounter, options={}) {
        if (!quickEncounter) {return;}
        let qeNote = null;
        //Create a Map Note for this encounter - the default is where the saved Tokens were
        let noteAnchor = {}
        if (quickEncounter.coords) {
            noteAnchor = {
                x: quickEncounter.coords.x,
                y: quickEncounter.coords.y
            }
        } else if (options.placeDefault) {
            //Otherwise, place it in the middle of the canvas stage (current view)
            noteAnchor = {
                x : canvas.stage.pivot.x,
                y : canvas.stage.pivot.y
            }
        } else {return;}
        // Validate the final position is in-bounds
        //1.0.4l: Use canvas.stage.hitArea in v10
        const hitArea = QuickEncounter.isFoundryV10Plus ? canvas.stage.hitArea : canvas.grid.hitArea;
        if (hitArea.contains(noteAnchor.x, noteAnchor.y)) {
            // Create a Note; we don't pop-up the Note sheet because we really want this Note to be placed
            //(they can always edit it afterwards)
            qeNote = await EncounterNote.create(quickEncounter, noteAnchor);
        }
        return qeNote;
    }

    static getEncounterScene(journalEntry) {
        if (!journalEntry) {return null;}
        //1.0.4k: Use parent (which is what is saved to the map) if this is JournalEntryPage
        //1.0.7a: Check for FoundryV10
        const parentJournalEntry = (QuickEncounter.isFoundryV10Plus && (journalEntry instanceof JournalEntryPage)) ? journalEntry.parent : journalEntry;
        //if sceneNote is available, then we're in the Note Scene already
        if (parentJournalEntry.sceneNote) {return game.scenes.viewed;}
        else {          
            //Now we need to search through the available scenes to find a note with this Journal Entry
            for (const scene of game.scenes) {
                let notes;
                if (QuickEncounter.isFoundryV10Plus) {
                    notes = scene.notes;
                } else {
                    notes = scene.data.notes;
                }
                let foundNote;
                if (QuickEncounter.isFoundryV10Plus) {
                    foundNote = Array.from(notes.values()).find(nd => nd.entryId === parentJournalEntry.id);
                } else if (QuickEncounter.isFoundryV8Plus) {
                    foundNote = Array.from(notes.values()).find(nd => nd.data.entryId === parentJournalEntry.id);
                } else {
                    foundNote = notes.find(note => note.entryId === parentJournalEntry.id);
                }
                if (foundNote) {
                    return scene;
                }
            }
        }
        return null;
    }

    static async switchToMapNoteScene(qeScene, qeJournalEntry) {
        if (!qeScene) {return null;}
        await qeScene.view();
        //bail out if the Map Note hasn't been placed after 2s
        let timer = null;
        for (let count=0; count<10; count++) {
            timer = setTimeout(() => {},200);
            if (qeJournalEntry.sceneNote) {break;}
        }
        clearTimeout(timer);
        return qeJournalEntry.sceneNote;
    }

    //1.0.2c: Changed to sync operation (only called if there is no map note when you run)
    static noMapNoteDialog(quickEncounter, event, options) {
        Dialog.confirm({
            title: game.i18n.localize("QE.NoMapNote.TITLE"),
            content : game.i18n.localize("QE.NoMapNote.CONTENT"),
            yes : () => {
                //1.1.5e: Place a default MapNote and then call run() again
                EncounterNote.place(quickEncounter, {placeDefault : true}).then(() => {
                    //New approach - just pop up a prompt dialog
                    EncounterNote.dialogPrompt({
                        title: game.i18n.localize("QE.CreatedMapNote.TITLE"),
                        content: game.i18n.localize("QE.CreatedMapNote.CONTENT"),
                        label: "",
                        options: {
                            top: window.innerHeight - 350,
                            left: window.innerWidth - 720,
                            width: 400,
                            jQuery: false
                        }
                    });
                });
                return true;
            }
        });
    }

    static async mapNoteIsPlaced(qeScene, journalEntry) {
        //Get the scene for this Quick Encounter (can't use sceneNote if we're in the wrong scene)
        if (!qeScene || !journalEntry) {return false;}
        //1.0.4k: Use parent (which is what is saved to the map) if this is JournalEntryPage
        //1.0.7a: Check for FoundryV10
        const parentJournalEntry = (QuickEncounter.isFoundryV10Plus && (journalEntry instanceof JournalEntryPage)) ? journalEntry.parent : journalEntry;
        //If we're viewing the relevant scene and the map note was placed, then good
        if (parentJournalEntry.sceneNote) {return true;}

        //Otherwise ask if you want to switch to the scene - default is No/false
        let shouldSwitch = false;
        //v0.6.12: Testing parameterization of i18n strings, using Localization.format()
        // If there is an 0612 version use that with a parameter, otherwise there isn't a parameter yet and we do it the pre-0.6.12 way
        let content; 
        if (game.i18n.has("QE.SwitchScene.CONTENT_v0612", false)) {
            content = game.i18n.format("QE.SwitchScene.CONTENT_v0612", {sceneName : qeScene.name});
        } else {
            content = game.i18n.localize("QE.SwitchScene.CONTENT") + qeScene.name + "?";
        }
        await Dialog.confirm({
            title: game.i18n.localize("QE.SwitchScene.TITLE"),
            content : content,
            //0.5.0 Need the Yes response to wait until we are in the correct scene (so don't make it async)
            //and in particular, the Journal Note has been drawn
            yes : () => {shouldSwitch = true},
            no : () => {shouldSwitch = false}
        });
        if (shouldSwitch) {return EncounterNote.switchToMapNoteScene(qeScene, parentJournalEntry);}
        else {return false;}
    }

    static async checkForInstantEncounter(quickEncounter, qeAnchor) {
        //already confirmed that the setting is there
        const dialogData = {
            title: game.i18n.localize("QE.CheckInstantEncounter.TITLE"),
            content : game.i18n.localize("QE.CheckInstantEncounter.CONTENT"),
            button1cb : (html, event) => {
                const options = {
                    isInstantEncounter : true,
                    qeAnchor: qeAnchor
                }
                quickEncounter.run(event, options) 
            },
            button2cb : () => EncounterNote.create(quickEncounter, qeAnchor),
            button3cb : null,
            buttonLabels :  [game.i18n.localize("QE.CheckInstantEncounter.BUTTON.RUN_INSTANT"),
                            game.i18n.localize("QE.CheckInstantEncounter.BUTTON.CREATE_QE"),""],
            options : {}
        }

        Dialog3.buttons3(dialogData);
    }//end checkForInstantEncounter()

    //1.1.5 Created to handle JE or JEPage for hook on dropCanvasData - recloned NotesLayer._onDropData()
    static async checkForQEAndCreateNote(journalEntry, data) {
        //Takes either a JournalEntry or JournalEntryPage that is being dropped onto the canvas
        let journalEntryOrJEPage = journalEntry;
        console.warn("Replaced Journal Entry drag-and-drop with QuickEncounter.EncounterNote handling");

        //JournalEntry: Check if this is from a Compendium and also for migration from pre-v10 Foundry
        let quickEncounter;
        if (journalEntryOrJEPage instanceof JournalEntry) {
            if (journalEntryOrJEPage.compendium ) {
                //If it's a Compendium JE, have to create one before dropping
                const journalData = game.journal.fromCompendium(journalEntryOrJEPage);
                journalEntryOrJEPage = JournalEntry.implementation.create(journalData);
            }

            quickEncounter = QuickEncounter.extractQuickEncounterFromJEOrEmbedded(journalEntryOrJEPage);
            //Pre-v10 QEs are dynamically moved onto JournalEntryPage0 so check for that
            if (!quickEncounter && QuickEncounter.isFoundryV10Plus) {
                //1.0.4m We have to check at least the first JournalEntryPage (we know that we have dropped a Journal Entry)
                const journalEntryPage0 = journalEntryOrJEPage.pages?.values().next().value;
                quickEncounter = QuickEncounter.extractQuickEncounterFromJEOrEmbedded(journalEntryPage0);
            }
        } else { //JournalEntryPage 
            quickEncounter = QuickEncounter.extractQuickEncounterFromJEOrEmbedded(journalEntryOrJEPage);
        }

        //Get the world-transformed drop position - fortunately these have already been placed in (data.x,data.y) by Canvas._onDrop
        const noteAnchor = {x: data?.x, y: data?.y}
        if (quickEncounter) {
            //Confirmed this is a Quick Encounter
            //If we're checking for Instant Encounters, then pop a dialog
            if (game.settings.get(QE.MODULE_NAME, "checkForInstantEncounter")) {
                EncounterNote.checkForInstantEncounter(quickEncounter, noteAnchor);
            } else {
                EncounterNote.create(quickEncounter, noteAnchor);
            }
        } else {
            //create a normal Journal Entry Note
            const noteData = {entryId: journalEntryOrJEPage.id, x: noteAnchor.x, y: noteAnchor.y}
            //Another hack - because we don't have event we recover the raw drop location
            const clientX = (noteAnchor.x * canvas.stage.scale.x) + canvas.notes.worldTransform.tx;
            const clientY = (noteAnchor.y * canvas.stage.scale.y) + canvas.notes.worldTransform.ty;
            return canvas.notes._createPreview(noteData, {top: clientY - 20, left: clientX + 40});
        }
    }//end async checkForQEAndCreateNote()
}

//Delete any corresponding Map Notes if you delete the Journal Entry
Hooks.on("deleteJournalEntry", EncounterNote.delete);

//Pretty up the first Map Note (hopefully we can do the same for others)
Hooks.on(`renderEncounterNoteConfig`, async (noteConfig, html, data) => {
    const updateEncounterMapNote = game.i18n.localize("QE.UpdateEncounterMapNote.BUTTON");
    html.find('button[name="submit"]').text(updateEncounterMapNote);
});

//1.0.1: Instant Encounters - intercept Note creation and check if it's an Instant Encounter (initially with a dialog)
//If you drag a Quick Encounter Journal Entry to the Scene, then intercept it to render it similarly,
//and also ask about Instant Encounters
//Note that intercepting preCreateNoteDocument is too late because we are approving the preview at that point
//and renderNoteConfig it's too difficult to change the form of the Note
Hooks.on(`dropCanvasData`, (canvas, data) => {
    //This is a hack because we're basically replicating canvas.notes._onDropData()
    // Acquire Journal entry 
    //- because it's async and this hook can't be (otherwise it prematurely returns true and creates a preview) use .then chaining
    //1.1.5 check for either JournalEntry OR JournalEntryPage
    if (data?.type === "JournalEntry") {
        JournalEntry.fromDropData(data).then(journalEntry => {
            EncounterNote.checkForQEAndCreateNote(journalEntry, data);
        });
    } else if (data?.type === "JournalEntryPage") {
        JournalEntryPage.fromDropData(data).then(journalEntryPage => {
            EncounterNote.checkForQEAndCreateNote(journalEntryPage, data);
        });
    } else {return true;}   //handle dropping something else
    return false;   //stop processing - we're replacing Journal Note creation entirely
});
